### ISSUE TITLE
pythreejs Low Level Modifications (e.g. add threeJS Spot_Light)

### ISSUE THREAD

**Issue by TimoFriedri** [maartenbreddels/ipyvolume#276]
m using your library extensively for the plotting of 3D objects (meshes and voxel models).

Unfortunately, the lightning is quite simple and it is hard to infer the shape from some angles. 
I'd like to add additional light sources and activate shadows, which seems to be possible from the pythreejs side.

So far I have not been successful is doing so. It is quite hard to figure out what I have to edit/add ... in the respective scenes/meshes/ ... which are stored in a figure.widgets container.

Could someone give some hints on how to achieve "adding and modifying lightning for ipyvolume plot?

Maybe an low level modification guide in the docs would be nice.

Thanks

**Comment by maartenbreddels** [comment]
Hi Timo,

I think we have to change the shaders to take into account the pythreejs lighting. That involves diving into threejs, I am happy to assist you with the ipyvolume side of things.
Basically `js\glsl\*`needs to be  edited, and Figure should probably be an Scene subclass, or Object3D, so it can have lights, and then follow something along these lines:
https://stackoverflow.com/questions/30151086/threejs-how-do-i-make-a-custom-shader-be-lit-by-the-scenes-lights.
Do you feel like attacking this issue?

cheers,

Maarten

**Comment by rinftech-github** [comment]
Greetings!

We are working on adding Lighting(Ambiental, Directional, Spot, Point, Hemisphere) support for ipyvolume rendering.
Below we are describing how we are planning to go about it.

Similar to pylab functions like plot_mesh, we are adding functions for every light type(exp addDirectionalLight()), each with specific parameters (some being adjsutable through sliders and color pickers).
Further, a Light widget class is necessary with associated TypeScript classes(LightView, LightModel).

Our current shader tests feature vertex and fragment code from three.js(Lambert, Phong, Physical), combined with existing ipyvolume shader code, in order to be up to the visual standard from three.js examples. 

What is your opinion regarding the plan above?
Are there any shader modules that should not be present in the new code? (exp: skinning, morphtargets)
We are planning to specify as parameter the target scene in which a specific light will be added in order to diferentiate between scene_scatter, scene_volume, scene_opaque. Do you agree with this approach?

Please, if you have time, can you give any feedback in order for us to better prepare for the eventual Pull Request?

**Comment by rinftech-github** [comment]
We are concentrating on this until the end of March 2020. So we would appreciate any feedback as soon as possible. Thank you!

**Comment by maartenbreddels** [comment]
That is awesome to hear. I think no widgets should be required, I think we should try to build as much as possible on top of pythreejs. What do think about that?

**Comment by rinftech-github** [comment]
Thank you for your answer! 
So you propose not to add classes Light(widgets.Widget) (python) and LightView, LightModel (ts)? 
In order to specify the scene in which the a light is added (scatter, volume, opaque) where do you recommend to add the logic?

**Comment by maartenbreddels** [comment]
Indeed, I thought a bit about this, and I think ipyvolume can use for instance:
 * https://pythreejs.readthedocs.io/en/stable/api/lights/AmbientLight_autogen.html

I did the same with the Camera class/widget. It should also be possible to rely on the shader snippets from pythreejs. I think it does not require a lot of coding, but a lot of sorting out how threejs works. But I think it will pay of in not having to make a lot of new widgets, and possibly have more integrations with other threejs based projects.

**Comment by rinftech-github** [comment]
I'm really grateful for your prompt answers!
I investigated the code more and you are right. Creating a separate Light widget would be an nonoptimal way to go about it. Using pythree.js would be much easier.
Below is a very simple test snippet added to pylab.py. addAmbientLight is called within a jupyter notebook.

 def addAmbientLight(color=default_color):
    ambientLight = pythreejs.AmbientLight(color)
    fig = gcf()
    fig.scene.add(ambientLight)

This does not give the desired result - no light added. Mention: I'm using plot_mesh, so scene_scatter should contain the added light(s). Shouldn't there be a link between this.renderer.scene_scatter(FigureView) and fig.scene ?

Again, thank you for your answers, help and time!

**Comment by rinftech-github** [comment]
@maartenbreddels Please excuse the long list of questions. Let me reduce them to just one:
Is there a way to access the three.js scene from python code? (pylab, widgets etc). ipv.gcf() does not appear to be it. Thank you!

**Comment by maartenbreddels** [comment]
No, there is no scene in the python side, it is not exposed, it's only on the js side.

**Comment by rinftech-github** [comment]
Thank you :)
I almost finished the implementation for adding all types of lights + shadow support(if available).

**Comment by maartenbreddels** [comment]
Ok, there is actually!
But indeed, it requires adding the code snippets to the shaders, are you saying you have that working?? :)

**Comment by rinftech-github** [comment]
Yes :)
https://imgur.com/a/wLsmg9x

**Comment by maartenbreddels** [comment]
Whow, I'm impressed! :)
Looking forward to seeing a PR. Please open a PR soon, so I can take a look, it doesn't have to be pretty even.

**Comment by rinftech-github** [comment]
Thank you! I will open a PR as soon as possible!

**Comment by rinftech-github** [comment]
Hello! Before I submit a PR, I want to test the new scatter js and shader code.
For whatever reason, when I run 'pip install -e .' the files scatter.ts and associated new shaders are not updated. I did not encounter this problem with mesh.ts + shaders.

**Comment by TimoFriedri** [comment]
Hi,

the PR will be handled from our official HRI-EU github account (https://github.com/HRI-EU).
I plan to manage this the next days/weeks.

The first implementations of light support should be finished tomorrow.

**Comment by TimoFriedri** [comment]
Hi,

here is some feedback from @rinftech-github regarding the issues remaining with the scatter plot type:

-------------------
Regarding the shadow mapping issue with instanced geometry, this was discovered in three.js back in 2018:
[](https://github.com/mrdoob/three.js/issues/13995)

After some back and forth the maintainer implemented a fix in September 2019 which looks like it was included in release 109 as seen here:
[](https://github.com/mrdoob/three.js/pull/17505)

To benefit from it we would need pyhtreejs to be updated to at least three.js 109; We opened an issue on their github for that:
[](https://github.com/jupyter-widgets/pythreejs/issues/319)

I tried to assess what it would look like for us to make the change and it is not something to be done in a day or two, it would be so much more efficient for the maintainer to do the update.
After updating pythreejs, we would still have to refactor the code in the scatter class to use mrdoob's new class InstancedMesh so the fix wouldn't be applied automatically so to speak, but it would be solvable.

Attached is a screenshot with the shadow mapping issue – you can clearly see that the map only features one cube shadow.

![instance shadow mapping issue](https://user-images.githubusercontent.com/19408741/78531687-2d097780-77e6-11ea-94e3-f6e0673ef564.PNG)

----------------

Maybe we will tackle this the next weeks.
